src/DynamoCoreWpf/Services/IconsService.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Drawing;
+using System.Linq;
+using System.Reflection;
+using System.Resources;
+using System.Text;
+using System.Windows;
+using System.Windows.Media.Imaging;
+using Dynamo.DSEngine;
+
+namespace Dynamo.Wpf.Services
+{
+    public class IconsService
+    {
+        private static Dictionary<Assembly, IconsWarehouse> warehouses =
+            new Dictionary<Assembly, IconsWarehouse>();
+
+        internal static IconsWarehouse GetForAssembly(string assemblyPath)
+        {
+            var libraryCustomization = LibraryCustomizationServices.GetForAssembly(assemblyPath);
+            if (libraryCustomization == null)
+                return null;
+
+            var assembly = libraryCustomization.Assembly;
+
+            if (!warehouses.ContainsKey(assembly))
+                warehouses[assembly] = new IconsWarehouse(assembly);
+
+            return warehouses[assembly];
+        }
+    }
+
+    public class IconsWarehouse
```

Comment:
`IconsWarehouse`, such a creative name :)

I would have gone with `IconWarehouse` and `IconService` (both `Icon` in singular form).

I personally don't have an opinion about this, just checking in with @pboyer to see if he has a more brilliant idea?